Pass instanceId to mongodb metadata backend

welansari opened the pull request:
`instanceId` will be used when generating versionIds.
This will ensure uniqueness of the versionID across
all Cloudserver pods in the cluster.

The `instanceId` is a combination of the short hash
contained in the name of the pods, which is unique
across all pods within the same deployment, and a
prefix for differenciating the different types of
deployments (internal vs external).

see. https://scality.atlassian.net/wiki/spaces/OS/pages/3225583692/VersionID+Collisions+in+Zenko

Related PR: https://github.com/scality/Arsenal/pull/2445

Issue: CLDSRV-665

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

williamlardier commented on lib/Config.js:
```diff
@@ -906,6 +906,15 @@ class Config extends EventEmitter {
+            // - A short hash (5 characters) that is unique only within the current ReplicaSet/Deployment.
+            // To avoid the risk of collision between internal and external Cloudservers we use the longer
+            // hash as a replicationGroupId.
+            this.replicationGroupId = parts.length > 2 ? parts[parts.length - 2] : parts.pop();
```
Actually we cannot use only the long: as this will be the same for all the pods in the same deployments.

My suggestion, given the 7 characters constraint, would be to use the short one plus the two first characters of the long one.
It guarantees the uniqueness for a deployment, and the odds of facing this with the internal cloudserver is infinitesimal...

welansari replied:
agreed

jonathan-gramain commented on tests/unit/Config.js:
```diff
@@ -810,4 +810,56 @@ describe('Config', () => {
+            const config = new ConfigObject();
+            assert.strictEqual(config.replicationGroupId, 'RG001');
+        });
+        it('should genrate a new replicationGroupId if not set in config', () => {
```
```suggestion
        it('should generate a new replicationGroupId if not set in config', () => {
```

jonathan-gramain commented on lib/Config.js:
```diff
@@ -906,6 +906,20 @@ class Config extends EventEmitter {
+                this.replicationGroupId = parts.pop();
+            }
         } else {
             this.replicationGroupId = 'RG001';
```
We may one day hit the same issue if we have a "CRR" configuration on Artesca that uses a different replication group ID, right? Could be worth ensuring that if we're in an Artesca setting (how?), we don't have a replication group ID configured, and maybe refuse to start if that's the case.

jonathan-gramain replied:
Checking if the metadata backend is of type "mongo" could be the appropriate way I think, since the issue is related to our generation of client-side versions in the mongo wrapper.

Commit:
Pass instanceId to the mongodb metadata backend

instanceId will be used when generating versionIds.
This will ensure uniqueness of the versionID across
all Cloudserver pods in the cluster.

The instanceId is a combination of the short hash
contained in the name of the pods, which is unique
across all pods within the same deployment, and a
prefix for differenciating the different types of
deployments (internal vs external).

see. https://scality.atlassian.net/wiki/spaces/OS/pages/3225583692/VersionID+Collisions+in+Zenko

Issue: CLDSRV-665

Commit:
run mongo functional tests with base62 versionId encoding

Cloudserver in Zenko always uses base62 encoding, so we enforce
it in the mongo tests.

Issue: CLDSRV-665

welansari commented:
Note for reviewers, i updated this PR following the changes in the TAD.
We no longer use the `replicationGroupId` as a local unique id but a new field called `instanceId`.

ghost commented:
@Kerkesni do we want to target 8.8? Some clients might not be ready to upgrade to the latest product version right away

bert-e commented:
# Branches have diverged



This pull request's source branch `bugfix/CLDSRV-665` has diverged from
`development/9.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/9.0` into `bugfix/CLDSRV-665`
* Rebase `bugfix/CLDSRV-665` onto `origin/development/9.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

welansari commented:
@williamlardier yes forgot about that part, thanks

williamlardier commented on lib/Config.js:
```diff
@@ -968,6 +968,29 @@ class Config extends EventEmitter {
+            // We use the last part of the pod name as the instanceId, prefixed with 'i' for internal
+            // or 'e' for external cloudserver.
+            // This allows having a unique identifier that will be used when generating versionIds
+            const instanceType = process.env.HOSTNAME?.includes('internal') ? 'i' : 'e';
```
The `?` is not needed as we already check above

```suggestion
            const instanceType = process.env.HOSTNAME.includes('internal') ? 'i' : 'e';
```

williamlardier reviewed (commented):
lgtm otherwise, but let's wait for the finalized Arsenal part

welansari commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/9.0/bugfix/CLDSRV-665` with contents from `bugfix/CLDSRV-665`
and `development/9.0`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/9.0/bugfix/CLDSRV-665 origin/development/9.0
 $ git merge origin/bugfix/CLDSRV-665
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/9.0/bugfix/CLDSRV-665
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-665` into
`development/8.8`
* [w/9.0/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.0...w/9.0/bugfix/CLDSRV-665) will be merged into `development/9.0` 
* [w/9.1/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.1...w/9.1/bugfix/CLDSRV-665) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

Commit:
update arsenal to include versionID changes

Issue: CLDSRV-665

bert-e commented:
# History mismatch



Merge commit #cfba4d355c05814b43f9a5dd8b3d12f479547b48 on the integration branch
`w/9.0/bugfix/CLDSRV-665` is merging a branch which is neither the current
branch `bugfix/CLDSRV-665` nor the development branch
`development/9.0`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-665` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

Commit:
bump version to 8.8.55

Issue: CLDSRV-665

welansari commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/9.0/bugfix/CLDSRV-665` with contents from `bugfix/CLDSRV-665`
and `development/9.0`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/9.0/bugfix/CLDSRV-665 origin/development/9.0
 $ git merge origin/bugfix/CLDSRV-665
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/9.0/bugfix/CLDSRV-665
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-665` into
`development/8.8`
* [w/9.0/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.0...w/9.0/bugfix/CLDSRV-665) will be merged into `development/9.0` 
* [w/9.1/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.1...w/9.1/bugfix/CLDSRV-665) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

bert-e commented:
# History mismatch



Merge commit #c631a5388e3e1e63574956850fb62bd8f706a5ba on the integration branch
`w/9.0/bugfix/CLDSRV-665` is merging a branch which is neither the current
branch `bugfix/CLDSRV-665` nor the development branch
`development/9.0`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-665` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

Kerkesni commented:
/reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**





*The following options are set:* **create_integration_branches**

bert-e commented:
# History mismatch



Merge commit #c631a5388e3e1e63574956850fb62bd8f706a5ba on the integration branch
`w/9.0/bugfix/CLDSRV-665` is merging a branch which is neither the current
branch `bugfix/CLDSRV-665` nor the development branch
`development/9.0`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-665` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

welansari commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/9.0/bugfix/CLDSRV-665` with contents from `bugfix/CLDSRV-665`
and `development/9.0`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/9.0/bugfix/CLDSRV-665 origin/development/9.0
 $ git merge origin/bugfix/CLDSRV-665
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/9.0/bugfix/CLDSRV-665
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-665` into
`development/8.8`
* [w/9.0/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.0...w/9.0/bugfix/CLDSRV-665) will be merged into `development/9.0` 
* [w/9.1/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.1...w/9.1/bugfix/CLDSRV-665) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

francoisferrand commented on lib/Config.js:
```diff
@@ -968,6 +968,29 @@ class Config extends EventEmitter {
+        } else {
+            this.instanceId = uuidv4().replace(/-/g, '').slice(0, 6);
+        }
+
```
it feels wrong to include this logic (tied very strongly to k8s implementation details) in here : this really breaks separation of concerns...

However, not sure if/how we can do so much better. It may be possible to [inject k8s pod fields](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/) into the pod, and possible use both `name` (cloudserver-7869f99955-nr984) and `computedName` (cloudserver-7869f99955) to reduce coupling, possibly using a ZKOP-defined initContainer....

....but would need much more investigation and dependency with ZKOP, and not sure we can really wait here.

→ Let's create a ticket to improve this, and (eventually) move this code out of cloudserver and into ZKOP
→ In the mean time, we should keep Cloudserver futureproof by allowing to explicitly set the instanceId:

```
   const instanceId = env.CLOUDSERVER_INSTANCE_ID || config.InstanceID;
   if (instanceId) {
      this.instanceId = instanceId;
   } else if (process.env.HOSTNAME) {
      ...
   } else {
      this.instanceId = uuidv4().replace(/-/g, '').slice(0, 6);
   }
```

welansari replied:
created CLDSRV-722

francoisferrand commented on .github/workflows/tests.yaml:
```diff
@@ -255,6 +255,7 @@ jobs:
       S3METADATA: mongodb
       S3KMS: file
       S3_LOCATION_FILE: /usr/src/app/tests/locationConfig/locationConfigTests.json
+      S3_VERSION_ID_ENCODING_TYPE: base62
```
should we keep this "unset" in mongo-v0, to exercise both behaviors?
(or actually just remove mongo-v0, as we always use v1 with mongo)

welansari replied:
The code base still supports both v0 and v1 so we should keep the tests. I made v0 tests run with hex versionIDs

francoisferrand commented on .github/docker/docker-compose.yaml:
```diff
@@ -45,6 +45,7 @@ services:
       - SCUBA_HEALTHCHECK_FREQUENCY
       - S3QUOTA
       - QUOTA_ENABLE_INFLIGHTS
+      - S3_VERSION_ID_ENCODING_TYPE
```
instead of setting this here, should we (can we?) set it automatically in config.js when backend is `mongo` ?

welansari replied:
This will need quite a rework as `S3_VERSION_ID_ENCODING_TYPE` is used by the versionID generation code inside of Arsenal, it's not a simple change in Cloudserver

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

bert-e commented:
# History mismatch



Merge commit #e1fb0290322cc75e24b8e90c8fd2a9d4e6cd4177 on the integration branch
`w/9.0/bugfix/CLDSRV-665` is merging a branch which is neither the current
branch `bugfix/CLDSRV-665` nor the development branch
`development/9.0`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-665` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

welansari commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/9.0/bugfix/CLDSRV-665` with contents from `bugfix/CLDSRV-665`
and `development/9.0`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/9.0/bugfix/CLDSRV-665 origin/development/9.0
 $ git merge origin/bugfix/CLDSRV-665
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/9.0/bugfix/CLDSRV-665
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-665` into
`development/8.8`
* [w/9.0/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.0...w/9.0/bugfix/CLDSRV-665) will be merged into `development/9.0` 
* [w/9.1/bugfix/CLDSRV-665](https://github.com/scality/cloudserver/compare/development/9.1...w/9.1/bugfix/CLDSRV-665) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.8`

* :heavy_check_mark: `development/9.0`

* :heavy_check_mark: `development/9.1`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`

* :heavy_check_mark: `development/9.0`

* :heavy_check_mark: `development/9.1`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue CLDSRV-665.

Goodbye kerkesni.
